nix-parser/src/parser/tokens/strings.rs

```diff
@@ -0,0 +1,297 @@
+                    .push_str(&format!("\n{}", this_line_leading_literal.0));
+                last_line_trailing_literal.2 = this_line_leading_literal.2;
+                total_literals.append(&mut literals);
+                total_interpolations.append(&mut interpolations);
+                (total_literals, total_interpolations)
+            },
+        )
+    };
+    let interpolations: Partial<_> = interpolations.into_iter().collect();
+    let mut expr: Partial<_> = {
+        let mut literals = literals.into_iter().map(|(s, start, end)| {
+            Expr::Literal(Literal::String(
+                s,
+                original_input
+                    .slice(start.offset..end.offset)
+                    .to_byte_span(),
+            ))
+        });
+        let first: Partial<_> = literals
+            .next()
+            .expect("there is at least one literal on each line")
+            .into();
+        interpolations.flat_map(move |interpolations| {
+            interpolations.into_iter().zip(literals).fold(
+                first,
+                |first, (interpolation, literal)| {
+                    first.map(|first| {
+                        let first = Expr::Binary(ExprBinary::new(
+                            BinaryOp::Add,
+                            Box::new(first),
+                            Box::new(interpolation),
+                            Default::default(),
+                        ));
+                        Expr::Binary(ExprBinary::new(
+                            BinaryOp::Add,
+                            Box::new(first),
+                            Box::new(literal),
+                            Default::default(),
+                        ))
+                    })
+                },
+            )
+        })
+    };
+    if no_delimiter {
+        expr.extend_errors(VerboseError {
+            errors: vec![(
+                input,
+                VerboseErrorKind::Context("unexpected EOF, expecting end of multiline string"),
+            )],
+        });
+        Ok((input, expr))
+    } else {
+        Ok((input, expr))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use nom::combinator::all_consuming;
+
+    use crate::ast::{tokens::Ident, Bind, BindSimple, ExprSet};
+    #[test]
+    fn it_works() {
+        let input = Span::new(
+            r#"''
+hello
+world
+''"#,
+        );
+        let (_, val) = all_consuming(single_quote_string)(input).unwrap();
+        let span = Span::new("").to_byte_span();
+        assert_eq!(
+            val,
+            Partial::from(Expr::Literal(Literal::String(
+                "hello\nworld\n".into(),
+                span,
+            )))
+        );
+
+        let input = Span::new(
+            r#"''
+  hello
+  world
+  ${ { a = 10; } }
+''"#,
+        );
+        let (_, val) = all_consuming(single_quote_string)(input).unwrap();
+        assert_eq!(
+            val,
+            Partial::from(Expr::Binary(ExprBinary::new(
```

Comment:
This reminds me that I need to improve the `nix!()` macro for producing more AST types from basic inline Nix expressions. :smile: This shouldn't block the PR of course, but you can try to produce string literals, integers, and identifier paths using the `nix_token!()` macro or `Expr`s using the `nix!()` token, both available at the crate root. There is no set nor bind support just yet, unfortunately, which makes it limited in usefulness.

Reply:
I will try it out in the next PR, I guess. That macro looks awesome!